tests/README.md

```diff
@@ -0,0 +1,120 @@
+```
+
+While the resulting step is returned, it needn't be processed to keep the network going, as its messages are automatically added to the queue.
+
+Instead of targeting a node in particular, the same input can be sent to all nodes:
+
+```rust
+net.broadcast_input(input).expect("algorithm failed");
+```
+
+### Cranking the network
+
+The network advances through the `crank()` function, on every call
+
+1. the adversary is given a chance to re-order the message queue,
+1. the next message in the queue is delivered to its destination node (if the node is non-faulty) or the adversary (if the node is faulty),
+1. all messages from the resulting step are queued again,
+1. and the resulting step (or error) is returned.
+
+If there were no messages to begin with, `None` is returned instead.
+
+Cranking can be done manually:
+
+```rust
+let step = net.crank()
+              .expect("expected at least one messages")
+              .expect("algorithm error");
+
+// Shorthand:
+let step = net.crank_expect();
+```
+
+For convenience, an iterator interface is also available:
+
+```rust
+for res in net {
+    let (node_id, step) = res.expect("algorithm error");
+    // ...
+}
+```
+
+This has the drawback that access to the network is not available between cranks, since it will be borrowed inside the for-loop. A common workaround is using a while loop instead:
+
+```rust
+while let Some(res) = net.crank() {
+    let (node_id, step) = res.expect("algorithm error");
+    // `net` can still be mutable borrowed here.
+}
+```
+
+### Inspecting the network
+
+In addition to the returned `Step`s, the network and nodes can be queried through various methods: `VirtualNet::{nodes, faulty_nodes, correct_nodes, get, get_mut}`.
+
+### Adversaries
+
+Adversaries can be introduced through the `.adversary` method on the constructor and are expected to implement the `net::adversary::Adversay` trait. Generic adversaries are available in the same module, while algorithm-specific ones should live next to each test case.
+
+```rust
+// Missing example.
+```
+
+### Tracing
+
+By default, all network tests write traces of every network message into logfiles, named `net-trace_*.txt`. Each log stores one message per line, in the format of `[SENDER] -> [RECEIVER]: MSG`.
+
+This behavior can be controlled using the `HBBFT_TEST_TRACE` environment variable; if set and equal to `0` or `false`, this functionality is disabled. Tracing is enabled by default.
+
+The `NetBuilder` allows hard-coding the trace setting, any value passed will override environment environment settings:
+
+```rust
+let net = NetBuilder(0..10)
+  .trace(false)   // Never log network messages.
+  // ...
+```
+
+### Checking outputs
+
+As a convenience, all nodes capture any generated output during operation for inspection. The following code fragment demonstrates how to use this to verify the end result:
+
+```rust
+let first = net.correct_nodes().nth(0).unwrap().outputs();
```

Comment:
To me, the sentence "demonstrates how to [...] verify the end result" sounds a bit vague. Maybe something like following would be more explicit:

```
The following code fragment demonstrates how to verify that all non-faulty nodes have output the same thing.
```